app/src/main/kotlin/com/simplemobiletools/musicplayer/helpers/MultiPlayer.kt

```diff
@@ -0,0 +1,281 @@
+package com.simplemobiletools.musicplayer.helpers
+
+import android.app.Application
+import android.content.IntentFilter
+import android.media.AudioAttributes
+import android.media.AudioManager
+import android.media.MediaPlayer
+import android.net.Uri
+import android.os.PowerManager
+import androidx.core.content.getSystemService
+import androidx.media.AudioAttributesCompat
+import androidx.media.AudioFocusRequestCompat
+import androidx.media.AudioManagerCompat
+import com.simplemobiletools.musicplayer.extensions.config
+import com.simplemobiletools.musicplayer.receivers.HeadsetPlugReceiver
+
+class MultiPlayer(private val app: Application, private val callbacks: PlaybackCallbacks) : MediaPlayer.OnErrorListener, MediaPlayer.OnCompletionListener {
+    private var mCurrentMediaPlayer = MediaPlayer()
+    private var mNextMediaPlayer: MediaPlayer? = null
+
+    var isInitialized: Boolean = false
+        private set
+
+    private var becomingNoisyReceiverRegistered = false
+    private val becomingNoisyReceiver = HeadsetPlugReceiver()
+    private val becomingNoisyReceiverIntentFilter = IntentFilter(AudioManager.ACTION_HEADSET_PLUG).apply {
+        addAction(AudioManager.ACTION_AUDIO_BECOMING_NOISY)
+    }
+
+    private val audioManager: AudioManager? = app.getSystemService()
+    private var isPausedByTransientLossOfFocus = false
+    private val audioFocusListener = AudioManager.OnAudioFocusChangeListener { focusChange ->
+        when (focusChange) {
+            AudioManager.AUDIOFOCUS_GAIN -> {
+                if (!isPlaying() && isPausedByTransientLossOfFocus) {
+                    start()
+                    callbacks.onPlayStateChanged()
+                    isPausedByTransientLossOfFocus = false
+                }
+                setVolume(Volume.NORMAL)
+            }
+            AudioManager.AUDIOFOCUS_LOSS -> {
+                pause()
+                callbacks.onPlayStateChanged()
+            }
+            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT -> {
+                val wasPlaying = isPlaying()
+                pause()
+                callbacks.onPlayStateChanged()
+                isPausedByTransientLossOfFocus = wasPlaying
+            }
+            AudioManager.AUDIOFOCUS_LOSS_TRANSIENT_CAN_DUCK -> {
+                setVolume(Volume.DUCK)
+            }
+        }
+    }
+
+    private val audioFocusRequest: AudioFocusRequestCompat =
+        AudioFocusRequestCompat.Builder(AudioManagerCompat.AUDIOFOCUS_GAIN).setOnAudioFocusChangeListener(audioFocusListener).setAudioAttributes(
```

Comment:
lets add brackets to this variable, theres too much code for a simple "="

Reply:
> add brackets to this variable

Do you mean to break it into multiple lines? [Check again, please.](https://github.com/SimpleMobileTools/Simple-Music-Player/pull/522/commits/96ad1973bf01b60bdc952e8e565634970e8eea17)

Reply:
yeah, change it to a "return ..." instead of replacing return with "="

Reply:
won't that lead to the AudioFocusRequest object being unnecessarily recreated every time we call `getAudioFocusRequest()` (i.e. every time we call abandonFocus() or requestFocust())?

we can instead make it nullable and move the initialization to init or another method

Reply:
okay, done